base_location/tests/test_base_location.py

```diff
@@ -0,0 +1,282 @@
+        """ Test onchange on res.better.zip """
+        usa_MA = self.env.ref('base.state_us_34')
+        better_zip1 = self.env['res.better.zip'].new(self.values_better_zip1())
+        better_zip1.state_id = usa_MA
+        better_zip1._onchange_state_id()
+        self.assertEqual(better_zip1.country_id, usa_MA.country_id)
+
+    def test_onchange_better_zip_city_id(self):
+        better_zip2 = self.env['res.better.zip'].new(self.values_better_zip2())
+        better_zip2.city_id = self.city_madrid
+        better_zip2._onchange_city_id()
+        self.assertEqual(better_zip2.city, self.city_madrid.name)
+
+    def test_onchange_better_zip_country_id(self):
+        better_zip1 = self.env['res.better.zip'].new(self.values_better_zip1())
+        better_zip1.country_id = self.env.ref('base.es')
+        better_zip1._onchange_country_id()
+        self.assertFalse(better_zip1.state_id)
+
+    def test_onchange_better_zip_none(self):
+        better_zip1 = self.env['res.better.zip'].new(self.values_better_zip1())
+        better_zip1.country_id = False
+        better_zip1._onchange_country_id()
+        self.assertFalse(better_zip1.state_id)
+
+    def test_onchange_partner_city_completion(self):
+        partner1 = self.env['res.partner'].new({
+            'name': 'Camptocamp',
+        })
+        better_zip2 = self.env['res.better.zip'].create(
+            self.values_better_zip2())
+        better_zip2.country_id.enforce_cities = True
+        partner1.zip_id = better_zip2
+        partner1._onchange_zip_id()
+        self.assertEqual(partner1.zip, better_zip2.name)
+        self.assertEqual(partner1.city, better_zip2.city)
+        self.assertEqual(partner1.state_id, better_zip2.state_id)
+        self.assertEqual(partner1.country_id, better_zip2.country_id)
+
+    def test_onchange_company_city_completion(self):
+        company = self.env['res.company'].new({'name': 'Test'})
+        better_zip1 = self.env['res.better.zip'].create(
+            self.values_better_zip1())
+        company.zip_id = better_zip1
+        company.on_change_city()
+        self.assertEqual(company.zip, better_zip1.name)
+        self.assertEqual(company.city, better_zip1.city)
+        self.assertEqual(company.state_id, better_zip1.state_id)
+        self.assertEqual(company.country_id, better_zip1.country_id)
+
+    def test_company_address_fields(self):
+        better_zip1 = self.env['res.better.zip'].create(
+            self.values_better_zip1()
+        )
+        partner = self.env['res.partner'].create({
+            'name': 'Test',
+            'zip_id': better_zip1.id,
+            'state_id': better_zip1.state_id.id,
+            'country_id': better_zip1.country_id.id,
+            'city_id': better_zip1.city_id.id,
+            'city': better_zip1.city,
+            'zip': better_zip1.name,
+        })
+        company = self.env['res.company'].new({
+            'name': partner.name,
+            'partner_id': partner.id
```

Comment:
This is not Odoo flow: you have to create the company, and the partner is created from it.